DIPs/DIP1005.md

```diff
@@ -0,0 +1,381 @@
+
+To generalize the observations above, fine-grained loading of modules is possible under either (or a combination of) the following circumstances: (a) the `static import` setup; (b) the selective `import` setup.
+
+Our assessment of this setup is as follows. The advantages are:
+
+* Fine-grained loading of imports is achieved with no changes in the language definition, only the implementation.
+* Project discipline may be enforced with relative ease, either manually or by means of simple tools. The rule is: "All private imports must be either `static import` or selective `import`".
+
+The disadvantages are:
+
+* The fine-grained _dependency structure_ is not attained by the selective `import` approach. A declaration using unspecified names does not clarify which `import`s it implicitly relies on. The relationship at distance remains between the `import` and the use thereof.
+* The `static import` setup does not share the issue above, at the cost of being cumbersome to use---all imported symbols must use full lookup everywhere. A reasonable engineering approach would be to define shorter names:
+
+```d
+    static import std.range.primitives;
+    alias isInputRange = std.range.primitives.isInputRange;
+    alias isForwardRange = std.range.primitives.isInputRange;
+    ...
+```
+
+    Such scaffolding is of course undesirable in the first place. Also, at least by the current language rules, such `alias` definitions would need to load the module anyway so as to ensure the name does exist. In order for this idiom to work, it would require subtle changes to the language that specify how certain `alias` declarations are exempt from early checking and delayed to the first actual use.
+* In either setup, `import`s are collapsed into their union, usually at the top of the module. Such lists grow out of sync with the actual code because during maintenance the programmer working on one declaration is not motivated to simultaneously alter a module-level `import` list shared by all declarations in the module. Over time, the `import`s grow into a superset of the actual depedencies used by the code, and do not reflect which declarations cause which imports even when accurate.
+* The "carrying" aspect is lost: any migration of a declaration to another module must be followed by awkwardly doing surgery on the `import` list of the receiving module. Again, the migration may leave unused `import`s in the module the declaration is taken from. The only recourse to keeping the `import` list in sync is special tooling or time-consuming discipline (search the module for uses, attempt recompilation).
+
+Although we consider introducing lazy imports an improvement over the current state of affairs, our assessment is that such a feature would fall short of truly allowing a project to rein in its dependency structure.
```

Comment:
I use the royal "we" throughout to mean "the author(s) of this document". It is standard for such documents and I might acquire a second author. I understand how it may cause confusion what with my double role, and will try to rephrase.